File: WoofIOS/SwiftPackages/Domain/UseCase/Sources/UseCase/Favorites/IsDogFavoriteUseCase.swift
import Combine
import DI
import Models
import Repositories

@Factory
public final class IsDogFavoriteUseCase {
    private let favoritesRepository: FavoritesRepository

    public func callAsFunction(dog: Dog) -> AnyPublisher<Bool, Never> {
        favoritesRepository.favoriteDogIds
            .map { $0.contains(dog.id) }
            .removeDuplicates()
            .eraseToAnyPublisher()
    }
}

File: WoofAndroid/domain/usecase/src/main/kotlin/com/perrystreet/woof/usecase/favorites/IsDogFavoriteUseCase.kt
package com.perrystreet.woof.usecase.favorites

import com.perrystreet.woof.models.dog.Dog
import com.perrystreet.woof.repositories.favorites.FavoritesRepository
import io.reactivex.rxjava3.core.Observable
import org.koin.core.annotation.Factory

@Factory
class IsDogFavoriteUseCase(
    private val favoritesRepository: FavoritesRepository,
) {
    operator fun invoke(dog: Dog): Observable<Boolean> =
        favoritesRepository.favoriteDogIds
            .map { dog.id in it }
            .distinctUntilChanged()
}
